### PR TITLE
LWE: migrate RLWE Type to NewLWEType

### PR DESCRIPTION
**Pull request by ZenithalHourlyRate**
Check #785 and #876.

Very WIP as we have RLWE Type everywhere. It also depends on on-going PRs (#1176 and #1183)

Currently get the `secret-to-bgv` working, should review how the type converter converts SecretType to a NewLWECiphertextType, as some fields are still ambiguous or without description.

```mlir
!Z1005037682689_i64_ = !mod_arith.int<1005037682689 : i64>
!Z1032955396097_i64_ = !mod_arith.int<1032955396097 : i64>
!Z1095233372161_i64_ = !mod_arith.int<1095233372161 : i64>
!Z65537_i64_ = !mod_arith.int<65537 : i64>
#full_crt_packing_encoding = #lwe.full_crt_packing_encoding<scaling_factor = 0>
#key = #lwe.key<>
#modulus_chain = #lwe.modulus_chain<elements = <1095233372161 : i64, 1032955396097 : i64, 1005037682689 : i64, 998595133441 : i64, 972824936449 : i64, 959939837953 : i64>, current = 2>
#modulus_chain1 = #lwe.modulus_chain<elements = <1095233372161 : i64, 1032955396097 : i64, 1005037682689 : i64, 998595133441 : i64, 972824936449 : i64, 959939837953 : i64>, current = 0>
#modulus_chain2 = #lwe.modulus_chain<elements = <1095233372161 : i64, 1032955396097 : i64, 1005037682689 : i64, 998595133441 : i64, 972824936449 : i64, 959939837953 : i64>, current = 1>
!rns_L0_ = !rns.rns<!Z1095233372161_i64_>
!rns_L1_ = !rns.rns<!Z1095233372161_i64_, !Z1032955396097_i64_>
!rns_L2_ = !rns.rns<!Z1095233372161_i64_, !Z1032955396097_i64_, !Z1005037682689_i64_>
#ring_Z65537_i64_1_x8_ = #polynomial.ring<coefficientType = !Z65537_i64_, polynomialModulus = <1 + x**8>>
#plaintext_space = #lwe.plaintext_space<ring = #ring_Z65537_i64_1_x8_, encoding = #full_crt_packing_encoding>
#ring_rns_L0_1_x8_ = #polynomial.ring<coefficientType = !rns_L0_, polynomialModulus = <1 + x**8>>
#ring_rns_L1_1_x8_ = #polynomial.ring<coefficientType = !rns_L1_, polynomialModulus = <1 + x**8>>
#ring_rns_L2_1_x8_ = #polynomial.ring<coefficientType = !rns_L2_, polynomialModulus = <1 + x**8>>
#ciphertext_space_L0_ = #lwe.ciphertext_space<ring = #ring_rns_L0_1_x8_, encryption_type = lsb>
#ciphertext_space_L1_ = #lwe.ciphertext_space<ring = #ring_rns_L1_1_x8_, encryption_type = lsb>
#ciphertext_space_L2_ = #lwe.ciphertext_space<ring = #ring_rns_L2_1_x8_, encryption_type = lsb>
#ciphertext_space_L2_D3_ = #lwe.ciphertext_space<ring = #ring_rns_L2_1_x8_, encryption_type = lsb, size = 3>
!rlwe_ct_L0_ = !lwe.new_lwe_ciphertext<application_data = <message_type = i16>, plaintext_space = #plaintext_space, ciphertext_space = #ciphertext_space_L0_, key = #key, modulus_chain = #modulus_chain1>
!rlwe_ct_L1_ = !lwe.new_lwe_ciphertext<application_data = <message_type = tensor<8xi16>>, plaintext_space = #plaintext_space, ciphertext_space = #ciphertext_space_L1_, key = #key, modulus_chain = #modulus_chain2>
!rlwe_ct_L1_1 = !lwe.new_lwe_ciphertext<application_data = <message_type = i16>, plaintext_space = #plaintext_space, ciphertext_space = #ciphertext_space_L1_, key = #key, modulus_chain = #modulus_chain2>
!rlwe_ct_L2_ = !lwe.new_lwe_ciphertext<application_data = <message_type = tensor<8xi16>>, plaintext_space = #plaintext_space, ciphertext_space = #ciphertext_space_L2_, key = #key, modulus_chain = #modulus_chain>
!rlwe_ct_L2_D3_ = !lwe.new_lwe_ciphertext<application_data = <message_type = tensor<8xi16>>, plaintext_space = #plaintext_space, ciphertext_space = #ciphertext_space_L2_D3_, key = #key, modulus_chain = #modulus_chain>
module {
  func.func @dot_product(%arg0: !rlwe_ct_L2_, %arg1: !rlwe_ct_L2_) -> !rlwe_ct_L0_ {
    %c1 = arith.constant 1 : index
    %c2 = arith.constant 2 : index
    %c4 = arith.constant 4 : index
    %c7 = arith.constant 7 : index
    %0 = bgv.mul %arg0, %arg1 : (!rlwe_ct_L2_, !rlwe_ct_L2_) -> !rlwe_ct_L2_D3_
    %1 = bgv.relinearize %0 {from_basis = array<i32: 0, 1, 2>, to_basis = array<i32: 0, 1>} : !rlwe_ct_L2_D3_ -> !rlwe_ct_L2_
    %2 = bgv.rotate %1 {offset = 4 : index} : !rlwe_ct_L2_
    %3 = bgv.add %1, %2 : !rlwe_ct_L2_
    %4 = bgv.rotate %3 {offset = 2 : index} : !rlwe_ct_L2_
    %5 = bgv.add %3, %4 : !rlwe_ct_L2_
    %6 = bgv.rotate %5 {offset = 1 : index} : !rlwe_ct_L2_
    %7 = bgv.add %5, %6 : !rlwe_ct_L2_
    %8 = bgv.modulus_switch %7 {to_ring = #ring_rns_L1_1_x8_} : !rlwe_ct_L2_ -> !rlwe_ct_L1_
    %9 = bgv.extract %8, %c7 : (!rlwe_ct_L1_, index) -> !rlwe_ct_L1_1
    %10 = bgv.modulus_switch %9 {to_ring = #ring_rns_L0_1_x8_} : !rlwe_ct_L1_1 -> !rlwe_ct_L0_
    return %10 : !rlwe_ct_L0_
  }
}


```

### CODE REVIEW

**Comment by AlexanderViand-Intel** [comment]
❤️

I guess this PR could also eventually address #858?

File: lib/Dialect/LWE/IR/NewLWEAttributes.td [hunk 1]
```diff
@@ -223,16 +223,8 @@ def LWE_FullCRTPackingEncoding
   }];
 }
 
-def LWE_AnyPlaintextEncodingAttr : LWE_EncodingAttrForLWE<"PlaintextEncoding", "plaintext_encoding"> {
```
Comment (AlexanderViand-Intel) [inline lib/Dialect/LWE/IR/NewLWEAttributes.td:226]: Nice simplification!

File: lib/Dialect/LWE/IR/LWEDialect.cpp [hunk 2]
```diff
@@ -421,6 +473,19 @@ LogicalResult verifyEncodingAndTypeMatch(mlir::Type type,
     return success();
   }
```
Comment (AlexanderViand-Intel) [inline lib/Dialect/LWE/IR/LWEDialect.cpp:475]: If the old encodings (`PolynomialEvaluationEncodingAttr` and `InverseCanonicalEmbeddingEncodingAttr`) have no more uses, can you just remove them?

File: tests/Dialect/BGV/Conversions/bgv_to_openfhe/bgv_to_openfhe.mlir [hunk 3]
```diff
@@ -1,33 +1,47 @@
...
+!rns_L0_ = !rns.rns<!Z1095233372161_i64_>
+!rns_L1_ = !rns.rns<!Z1095233372161_i64_, !Z1032955396097_i64_>
 
-#params1 = #lwe.rlwe_params<dimension=2, ring=#ring1>
-#params2 = #lwe.rlwe_params<dimension=4, ring=#ring1>
-#params3 = #lwe.rlwe_params<dimension=2, ring=#ring2>
-#params4 = #lwe.rlwe_params<dimension=3, ring=#ring1>
+#ring_Z65537_i64_1_x1024_ = #polynomial.ring<coefficientType = !Z65537_i64_, polynomialModulus = <1 + x**1024>>
+#ring_rns_L0_1_x1024_ = #polynomial.ring<coefficientType = !rns_L0_, polynomialModulus = <1 + x**1024>>
+#ring_rns_L1_1_x1024_ = #polynomial.ring<coefficientType = !rns_L1_, polynomialModulus = <1 + x**1024>>
```
Comment (j2kun) [inline tests/Dialect/BGV/Conversions/bgv_to_openfhe/bgv_to_openfhe.mlir:12]: File a followup issue for verifier of modulus chain level with actual underlying Polynomial coefficient type

File: lib/Dialect/LWE/IR/LWEDialect.cpp (hunk 2 above)
Comment (AlexanderViand-Intel) [reply lib/Dialect/LWE/IR/LWEDialect.cpp:475]: (can also wait until a follow-up PR that cleans up all the old stuff)

**Review by j2kun — APPROVED** [review]

**Pull request merged**